File: contact_moblie_client/lib/model/message.dart
import 'package:json_annotation/json_annotation.dart';

part 'message.g.dart';

@JsonSerializable()
class Message {
  /// 雪花ID */
  int? seqId;

  /// 服务器接受时间 */
  int? createdAt;

  /// 是否 发送到服务器 */
  bool? sync;
  String uuid;

  /// Snowflake long */
  int? conversationId;

  /// message from, setting by server */
  int? from;

  /// message send to */
  int? to;

  /// Receiver type */
  String type;

  /// Creator type */
  String creatorType;
  Content content;
  String? nickName;

  Message(
      {this.seqId,
      this.createdAt,
      this.sync = false,
      required this.uuid,
      this.conversationId,
      this.from,
      this.to,
      required this.type,
      required this.creatorType,
      required this.content,
      this.nickName});

  factory Message.fromJson(Map<String, dynamic> json) =>
      _$MessageFromJson(json);
  Map<String, dynamic> toJson() => _$MessageToJson(this);
}

@JsonSerializable()
class Content {
  String? contentType;
  String? sysCode;
  String? serviceContent;
  TextContent? textContent;
  PhotoContent? photoContent;
  // 后面再添加附件
  dynamic attachments;

  Content(
      {this.contentType,
      this.sysCode,
      this.serviceContent,
      this.textContent,
      this.photoContent,
      this.attachments});

  factory Content.fromJson(Map<String, dynamic> json) =>
      _$ContentFromJson(json);
  Map<String, dynamic> toJson() => _$ContentToJson(this);
}

@JsonSerializable()
class TextContent {
  String text;

  TextContent({required this.text});

  factory TextContent.fromJson(Map<String, dynamic> json) =>
      _$TextContentFromJson(json);
  Map<String, dynamic> toJson() => _$TextContentToJson(this);
}

@JsonSerializable()
class PhotoContent {
  String mediaId;
  String filename;
  int picSize;
  String type;

  PhotoContent({
    required this.mediaId,
    required this.filename,
    this.picSize = 0,
    required this.type,
  });

  factory PhotoContent.fromJson(Map<String, dynamic> json) =>
      _$PhotoContentFromJson(json);
  Map<String, dynamic> toJson() => _$PhotoContentToJson(this);
}

File: src/renderer/domain/Message.ts
import {
  CreatorType,
  MessageStatusKey,
  MessageTypeKey,
  SysCode,
} from './constant/Message';

/**
 * 标记为本地消息
 */
export interface LocalMessage {
  localType?: MessageTypeKey;
  status?: MessageStatusKey;
  file?: File;
  fileId?: string;
}

/**
 * Chat messages are not stored locally, all get from the server
 * Sync messages by redis, Persistence by ElasticSearch
 */
export interface Message extends MessageResponse, LocalMessage {
  uuid: string;
  /** Snowflake long */
  conversationId?: number;
  /** message from, setting by server */
  from?: number;
  /** message send to */
  to?: number;
  /** Receiver type */
  type: CreatorType;
  /** Creator type */
  creatorType: CreatorType;
  content: Content;
  nickName?: string;
}

export interface UpdateMessage {
  pts: number;
  message: Message;
  ptsCount: number;
}

export interface TextContent {
  text: string;
}

export interface PhotoContent {
  /** 媒体id */
  mediaId: string;
  /** 图片名称 */
  filename: string;
  /** 图片大小 */
  picSize: number;
  /** 图片类型 */
  type: string;
}

export interface Attachments {
  /** 媒体id */
  mediaId: string;
  /** 文件名称 */
  filename: string;
  /** 文件大小 */
  size: number;
  /** 根据类型显示不同图标 */
  type: string;
}

export interface Content {
  contentType: MessageTypeKey;
  /** sys 服务消息类型 * */
  sysCode?: SysCode;
  /** 服务消息内容， JSON方式 * */
  serviceContent?: string;
  /** 文字 */
  textContent?: TextContent;
  /** 图片 */
  photoContent?: PhotoContent;
  /** 附件 */
  attachments?: Attachments;
}
/** Message 去重 */
export interface MessagesMap {
  [uuid: string]: Message;
}

export interface MessageResponse {
  /** 雪花ID */
  seqId?: number;
  /** 服务器接受时间 */
  createdAt?: number | string | Date;
}
